### PR TITLE
perf: compress deserialization wasm stack by exploiting RTS stack for all deserialization args

### PR DESCRIPTION
**Pull request by crusso**
* introduces a traditional frame pointer to RTS stack  (distinct from Wasm stack) for storing arguments at known offset from frame pointer
* uses it to offload all args from Wasm to RTS stack, thus compressing the Wasm stack
* reaches parity in serialization/deserialization of deeply nested data (achievement unlocked). Well, if you call that deep.

Not as bad as red black trees.

- [x] reference previous values of locals via saved frame pointer, avoid spills to Wasm stack (also just cooler).
- [x] make expensive stack measurement code conditional on compiler flag.

Builds on #3798 and #3794.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
   * Performance improvement: heap allocator speedup (#3090, #3790).
 
+>>>>>>> origin/master
```
Comment (crusso) [inline Changelog.md:12]: yuck

File: Changelog.md [hunk 2]
```diff
@@ -1,12 +1,17 @@
 # Motoko compiler changelog
 
+* Improve recursive deserialization capacity to match recursive serialization capacity by reducing Wasm stack consumption (#3809).
```
Comment (crusso) [inline Changelog.md:3]:
```suggestion
* Improve recursive deserialization capacity to match recursive serialization capacity by reducing
  Wasm stack consumption (#3809).
```

File: Changelog.md [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
+  Think "shallow trees good, very long lists bad".
+
 ## 0.8.2 (2023-02-17)
 
 * motoko (`moc`)
 
   * Add compiler flag `--rts-stack-pages <n>` to override default number of
     pages dedicated to fixed runtime system stack. Now defaults to 32 pages
-    (2MiB) (up from previous 2 pages/128KiB) (#3782).
+    (2MiB) (up from previous 2 pages/128KiB) (#3782).CH
```
Comment (crusso) [inline Changelog.md:15]:
```suggestion
    (2MiB) (up from previous 2 pages/128KiB) (#3782).
```

File: src/codegen/compile.ml [hunk 4]
```diff
@@ -6307,19 +6374,26 @@ module MakeSerialization (Strm : Stream) = struct
...
-               get_typtbl_ptr ^^ load_unskewed_ptr ^^ Registers.set_typtbl env ^^
-               get_maintyps_ptr ^^ load_unskewed_ptr ^^ Registers.set_typtbl_end env ^^
-               get_typtbl_size_ptr ^^ load_unskewed_ptr ^^ Registers.set_typtbl_size env
-             end ^^
-             ReadBuf.read_sleb128 env get_main_typs_buf ^^
-             compile_unboxed_const 0l ^^ (* initial depth *)
-             can_recover ^^
-             deserialize_go env t ^^ set_val ^^
+              begin
+                (* set up register arguments *)
```
Comment (crusso) [inline src/codegen/compile.ml:6378]:
```suggestion
                (* set up invariant register arguments *)
```

File: src/codegen/compile.ml [hunk 5]
```diff
@@ -6307,19 +6374,26 @@ module MakeSerialization (Strm : Stream) = struct
...
+                get_rel_buf_opt ^^ Registers.set_rel_buf_opt env ^^
+                get_data_buf ^^ Registers.set_data_buf env ^^
+                get_ref_buf ^^ Registers.set_ref_buf env ^^
+                get_typtbl_ptr ^^ load_unskewed_ptr ^^ Registers.set_typtbl env ^^
+                get_maintyps_ptr ^^ load_unskewed_ptr ^^ Registers.set_typtbl_end env ^^
+                get_typtbl_size_ptr ^^ load_unskewed_ptr ^^ Registers.set_typtbl_size env
+              end ^^
+              (* set up frame arguments *)
```
Comment (crusso) [inline src/codegen/compile.ml:6386]:
```suggestion
              (* set up variable frame arguments *)
```

File: src/codegen/compile.ml [hunk 6]
```diff
@@ -1263,6 +1270,65 @@ module Stack = struct
...
+    alloc_words env (Int32.add n 1l) ^^
+    (* store the current frame_ptr at offset 0*)
+    get_frame_ptr env ^^
+    G.i (Store {ty = I32Type; align = 2; offset = 0l; sz = None}) ^^
+    get_stack_ptr env ^^
+    (* set_frame_ptr to stack_ptr *)
+    set_frame_ptr env ^^
+    (* do as f *)
+    f () ^^
+    (* assert frame_ptr == stack_ptr *)
```
Comment (crusso) [inline src/codegen/compile.ml:1297]:
```suggestion
    (* Sanity: assert frame_ptr == stack_ptr *)
```
Should I disable that in production?

File: src/codegen/compile.ml [hunk 7]
```diff
@@ -5600,17 +5667,23 @@ module MakeSerialization (Strm : Stream) = struct
...
-        deserialize_go env t
+        (* assumes idltyp on stack *)
+        Stack.with_frame env "frame_ptr" 3l (fun () ->
+          Stack.set_local env 0l ^^ (* arg get_idltyp *)
+            (* set up frame arguments *)
+            ( (* Reset depth counter if we made progress *)
+              ReadBuf.get_ptr get_data_buf ^^ get_old_pos ^^
+              G.i (Compare (Wasm.Values.I32 I32Op.Eq)) ^^
+              G.if1 I32Type
+                (Stack.get_prev_local env 1l ^^ compile_add_const 1l)
```
Comment (luc-blaeser) [inline src/codegen/compile.ml:5678]: Detail: Maybe the frame offsets of the specific local variables could be defined as constants.

File: test/run-drun/ok/max-stack-variant.drun-run.ok [hunk 8]
```diff
@@ -1,81 +1,87 @@
...
+debug.print: {bytes = 27_676; length = 27_648; stack = 1_437_772; stack_pages = 22}
+debug.print: {bytes = 28_700; length = 28_672; stack = 1_491_020; stack_pages = 23}
+debug.print: {bytes = 29_724; length = 29_696; stack = 1_544_268; stack_pages = 24}
+debug.print: {bytes = 30_748; length = 30_720; stack = 1_597_516; stack_pages = 25}
+debug.print: {bytes = 31_772; length = 31_744; stack = 1_650_764; stack_pages = 26}
+debug.print: {bytes = 32_796; length = 32_768; stack = 1_704_012; stack_pages = 27}
+debug.print: {bytes = 33_820; length = 33_792; stack = 1_757_260; stack_pages = 27}
+debug.print: {bytes = 34_844; length = 34_816; stack = 1_810_508; stack_pages = 28}
+debug.print: {bytes = 35_868; length = 35_840; stack = 1_863_756; stack_pages = 29}
 debug.print: IC0502: Canister rwlgt-iiaaa-aaaaa-aaaaa-cai trapped: stack overflow
```
Comment (luc-blaeser) [inline test/run-drun/ok/max-stack-variant.drun-run.ok:82]: Great!

**Review by luc-blaeser — APPROVED** [review]
Great improvement! And excellent idea of moving the local state to stack frames on RTS stack.

File: src/codegen/compile.ml (hunk 7 above)
Comment (crusso) [reply src/codegen/compile.ml:5679]: Good idea.

File: src/codegen/compile.ml [hunk 9]
```diff
@@ -4956,7 +5023,7 @@ module MakeSerialization (Strm : Stream) = struct
       E.add_global32 env "@@ref_buf" Mutable 0l;
       E.add_global32 env "@@typtbl" Mutable 0l;
       E.add_global32 env "@@typtbl_end" Mutable 0l;
-      E.add_global32 env "@@typtbl_size" Mutable 0l)
+      E.add_global32 env "@@typtbl_size" Mutable 0l)
```
Comment (ggreif) [inline src/codegen/compile.ml:5026]: stray space

Comment (crusso) [inline src/codegen/compile.ml:5026]:
```suggestion
      E.add_global32 env "@@typtbl_size" Mutable 0l)
```

File: src/codegen/compile.ml (hunk 7 above)
Comment (crusso) [reply src/codegen/compile.ml]: Done

File: src/codegen/compile.ml (hunk 9 above)
Comment (crusso) [reply src/codegen/compile.ml]: Fixed

**Pull request merged**